resteasy-jaxrs/src/main/java/org/jboss/resteasy/tracing/RESTEasyTracingLoggerImpl.java

```diff
@@ -0,0 +1,145 @@
+                    break;
+                case TRACE:
+                    loggingLevel = Logger.Level.DEBUG;
+                    break;
+                case VERBOSE:
+                    loggingLevel = Logger.Level.TRACE;
+                    break;
+                default:
+                    loggingLevel = Logger.Level.INFO;
+            }
+            if (logger.isEnabled(loggingLevel)) {
+                logger.log(loggingLevel,
+                        event.name() + ' ' + message.toString() + " [" + RESTEasyTracingInfo.formatDuration(duration) + " ms]");
+            }
+        }
+    }
+
+    private static void formatInstance(final Object instance, final StringBuilder text) {
+        text.append(instance.getClass().getName()).append(" @")
+                .append(Integer.toHexString(System.identityHashCode(instance)));
+    }
+
+    private static void formatResponse(final Response response, final StringBuilder text) {
+        text.append(" <").append(formatStatusInfo(response.getStatusInfo())).append('|');
+        if (response.hasEntity()) {
+            formatInstance(response.getEntity(), text);
+        } else {
+            text.append("-no-entity-");
+        }
+        text.append('>');
+    }
+
+    private static String formatStatusInfo(final Response.StatusType statusInfo) {
+        return String.valueOf(statusInfo.getStatusCode()) + '/' + statusInfo.getFamily() + '|' + statusInfo.getReasonPhrase();
+    }
+
+    private String formatInstance(Object instance) {
+        final StringBuilder text = new StringBuilder();
+        if (instance == null) {
+            text.append("null");
+        } else if ((instance instanceof Number) || (instance instanceof String) || (instance instanceof Method)) {
+            text.append(instance.toString());
+        } else if (instance instanceof Response.StatusType) {
+            text.append(formatStatusInfo((Response.StatusType) instance));
+        } else {
+            text.append('[');
+            formatInstance(instance, text);
+            if (instance.getClass().isAnnotationPresent(Priority.class)) {
+                text.append(" #").append(instance.getClass().getAnnotation(Priority.class).value());
+            }
+            if (instance instanceof WebApplicationException) {
+                formatResponse(((WebApplicationException) instance).getResponse(), text);
+            } else if (instance instanceof Response) {
+                formatResponse(((Response) instance), text);
+            }
+            text.append(']');
+        }
+        return text.toString();
+    }
+
+    @Override
+    public long timestamp(final RESTEasyTracingEvent event) {
+        if (isEnabled(event.level())) {
+            return System.nanoTime();
+        }
+        return -1;
+    }
+
+    @Override
+    public void flush(final MultivaluedMap<String, Object> headers) {
+        final String[] messages = tracingInfo.getMessages();
```

Comment:
Do we need any check here on in the callers of the flush method here to ensure the processing is performed only when needed? Or maybe we need this in any case? I'm just thinking about performances.

Reply:
@asoldano `flush()` method flushes the data in `tracingInfo` and put it into response headers, and `tracingInfo` is injected by `logImpl` method of `RESTEasyTracingLoggerImpl`:

```
tracingInfo.addMessage(message);
```

This is the single place that `tracingInfo` is added, and `tracingInfo` itself is private.

So the checking point is in `logImpl(...)` method:

```java
if (isEnabled(event.level())) {
...
    final RESTEasyTracingMessage message = new RESTEasyTracingMessage(event, duration, messageArgsStr);
    tracingInfo.addMessage(message);
...
```

So to use the tracingLogger, taking the usages in `SynchronousDispatcher` for example:

```java
...
tracingLogger.log(RESTEasyServerTracingEvent.FINISHED, response.getStatus());
...
tracingLogger.flush(response.getOutputHeaders());
...
```

tracingLogger will always call `log(...)` for server side tracing *and* add the info into `tracingInfo` according to the level set and mode set. This is controlled by `log(...)` method internally.

So if the level is to disable the tracing, then `tracingInfo` will be empty.

In conclusion:

- `tracingInfo` is manipulated insided `logImpl(...)` function and it's private data of `RESTEasyTracingLoggerImpl`.
- `logImpl(...)` takes care of the tracing threshold and tracing mode.
- `tracingInfo` is empty, if tracing level is set to OFF or when controlling by ON_DEMAND mode.
- `flush(...)` method will do nothing when `tracingInfo` is empty.

Reply:
OK, makes sense, thanks.